### ISSUE TITLE
ID3V2 TCON field always Latin1 encoded

### ISSUE THREAD

**Issue by fauxpark** [taglib/taglib#1259]
I discovered when trying to change the genre field encoding in Kid3 from Latin1 to UTF16 that it just doesn't take, and always reverts back to Latin1, regardless of what my default encoding setting was.

Initially my thought was that this might be intentional due to the rather conflicting wording of the spec - it states that the TCON frame should be a numeric string, which earlier is said to always be Latin1 encoded, but also allows plaintext genre strings (I think in practice this is the most common use).

So I asked @ufleisch about it, who could repro with the taglib plugin but not id3lib, and tracked it down to the following line here:

https://github.com/taglib/taglib/blob/225c73e1816c3200c67703557972586e287fa665/taglib/mpeg/id3v2/id3v2tag.cpp#L656

**Comment by sbooth** [comment]
Does the string you're setting contain any characters that are not Latin-1?

It seems that upon rendering, `TextIdentificationFrame` calls `Frame::checkTextEncoding()`, which in turn calls `isLatin1()` on each field. Unless `isLatin1()` returns false the frame uses Latin-1 regardless of which encoding was specified.

https://github.com/taglib/taglib/blob/225c73e1816c3200c67703557972586e287fa665/taglib/mpeg/id3v2/frames/textidentificationframe.cpp#L275

**Comment by fauxpark** [comment]
Sorry, I forgot some extra context. It does get forced to UTF16 if I add a non-Latin1 character. However the behaviour I'm seeing is not the case for other fields eg. title, artist, album, where I can freely change the encoding even if they only have Latin1 characters. That coupled with what the spec says (confusing as it is) about the TCON frame is what made me wonder whether or not it's expected behaviour.

**Comment by sbooth** [comment]
I think it is probably the expected behavior inferred from the comment:

https://github.com/taglib/taglib/blob/225c73e1816c3200c67703557972586e287fa665/taglib/mpeg/id3v2/frames/textidentificationframe.h#L100

Whether it makes the most sense is certainly debatable.

**Comment by fauxpark** [comment]
If that is the case then why is it not for the other fields I mentioned? And then what is the point in supposedly being able to force the encoding in Kid3?

**Comment by fauxpark** [comment]
I'm not sure I follow. To be clear, all the files I'm working with are 2.3, so UTF8 shouldn't come into the picture, unless taglib just uses it internally?

**Comment by sbooth** [comment]
I'd assumed you were using ID3v2.4, but the code in `Frame::checkTextEncoding()` will correctly use UTF16 even if the frame is created with UTF8.

**Comment by fauxpark** [comment]
Ah, I think I understand now. So with that change the `TCON` frame will be set to UTF8 internally, but before being written to the file it goes through `checkTextEncoding()` which I can see will recommend UTF16 instead. Since currently it is set to Latin1 the recommendation always ends up being for Latin1.

**Comment by ufleisch** [comment]
I don't think that there is really a reason for the current behavior besides it being simple (after all the commit was titled "Simple ID3v2.4 => ID3v2.3 frame migration"). I would suggest that we only downgrade ID3v2.4 frames to ID3v2.3 frames with ISO-8859-1 encoding if the original ID3v2.4 frame is also ISO-8859-1 encoded, otherwise we might not be able to encode all characters correctly (in the case of the date/year frames this is not very likely, but when converting TIPL to IPLS, there could be some performer having a non-ASCII name). What about the following code:

```
diff --git a/taglib/mpeg/id3v2/id3v2tag.cpp b/taglib/mpeg/id3v2/id3v2tag.cpp
index f97be59e..67787edf 100644
--- a/taglib/mpeg/id3v2/id3v2tag.cpp
+++ b/taglib/mpeg/id3v2/id3v2tag.cpp
@@ -56,6 +56,14 @@ namespace
 
   constexpr long MinPaddingSize = 1024;
   constexpr long MaxPaddingSize = 1024 * 1024;
+
+  /*!
+   * Downgrade ID3v2.4 text \a encoding to value supported by ID3v2.3.
+   */
+  String::Type downgradeTextEncoding(String::Type encoding)
+  {
+    return encoding == String::Latin1 ? String::Latin1 : String::UTF16;
+  }
 }  // namespace
 
 class ID3v2::Tag::TagPrivate
@@ -578,7 +586,8 @@ void ID3v2::Tag::downgradeFrames(FrameList *frames, FrameList *newFrames) const
     String content = frameTDOR->toString();
 
     if(content.size() >= 4) {
-      auto frameTORY = new ID3v2::TextIdentificationFrame("TORY", String::Latin1);
+      auto frameTORY = new ID3v2::TextIdentificationFrame(
+        "TORY", downgradeTextEncoding(frameTDOR->textEncoding()));
       frameTORY->setText(content.substr(0, 4));
       frames->append(frameTORY);
       newFrames->append(frameTORY);
@@ -587,17 +596,20 @@ void ID3v2::Tag::downgradeFrames(FrameList *frames, FrameList *newFrames) const
 
   if(frameTDRC) {
     if(String content = frameTDRC->toString(); content.size() >= 4) {
-      auto frameTYER = new ID3v2::TextIdentificationFrame("TYER", String::Latin1);
+      auto frameTYER = new ID3v2::TextIdentificationFrame(
+        "TYER", downgradeTextEncoding(frameTDRC->textEncoding()));
       frameTYER->setText(content.substr(0, 4));
       frames->append(frameTYER);
       newFrames->append(frameTYER);
       if(content.size() >= 10 && content[4] == '-' && content[7] == '-') {
-        auto frameTDAT = new ID3v2::TextIdentificationFrame("TDAT", String::Latin1);
+        auto frameTDAT = new ID3v2::TextIdentificationFrame(
+          "TDAT", downgradeTextEncoding(frameTDRC->textEncoding()));
         frameTDAT->setText(content.substr(8, 2) + content.substr(5, 2));
         frames->append(frameTDAT);
         newFrames->append(frameTDAT);
         if(content.size() >= 16 && content[10] == 'T' && content[13] == ':') {
-          auto frameTIME = new ID3v2::TextIdentificationFrame("TIME", String::Latin1);
+          auto frameTIME = new ID3v2::TextIdentificationFrame(
+            "TIME", downgradeTextEncoding(frameTDRC->textEncoding()));
           frameTIME->setText(content.substr(11, 2) + content.substr(14, 2));
           frames->append(frameTIME);
           newFrames->append(frameTIME);
@@ -607,7 +619,9 @@ void ID3v2::Tag::downgradeFrames(FrameList *frames, FrameList *newFrames) const
   }
 
   if(frameTIPL || frameTMCL) {
-    auto frameIPLS = new ID3v2::TextIdentificationFrame("IPLS", String::Latin1);
+    auto frameIPLS = new ID3v2::TextIdentificationFrame(
+      "IPLS", downgradeTextEncoding(frameTIPL ? frameTIPL->textEncoding()
+                                              : frameTMCL->textEncoding()));
 
     StringList people;
 
@@ -653,7 +667,8 @@ void ID3v2::Tag::downgradeFrames(FrameList *frames, FrameList *newFrames) const
     if(!genreText.isEmpty())
       combined += genreText;
 
-    frameTCON = new ID3v2::TextIdentificationFrame("TCON", String::Latin1);
+    frameTCON = new ID3v2::TextIdentificationFrame(
+      "TCON", downgradeTextEncoding(frameTCON->textEncoding()));
     frameTCON->setText(combined);
     frames->append(frameTCON);
     newFrames->append(frameTCON);
```

**Comment by ufleisch** [comment]
I have improved the IPLS case and submitted PR #1260.

**Comment by ufleisch** [comment]
Fixed with #1260.
